pkg/node/local_node_store.go

```diff
@@ -107,40 +99,28 @@ func NewLocalNodeStore(params LocalNodeStoreParams) (LocalNodeStore, error) {
 			s.complete = nil
 			s.emit = nil
 			s.mu.Unlock()
+
+			localNode = nil
 			return nil
 		},
 	})
 
 	return s, nil
 }
 
-// defaultLocalNodeStore constructs the default instance for the LocalNodeStore used by
-// address.go.
-func defaultLocalNodeStore() LocalNodeStore {
-	src, emit, complete := stream.Multicast[LocalNode](stream.EmitLatest)
-	s := &localNodeStore{
-		Observable: src,
-		valid:      true,
-		emit:       emit,
-		complete:   complete,
-	}
-	s.cond = sync.NewCond(&s.mu)
-	return s
-}
-
-func (s *localNodeStore) Get() LocalNode {
+// Get retrieves the current local node. Use Get() only for inspecting the state,
+// e.g. in API handlers. Do not assume the value does not change over time.
+// Blocks until the store has been initialized.
+func (s *LocalNodeStore) Get() LocalNode {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-
-	// Block until the value has been initialized.
-	for !s.valid {
-		s.cond.Wait()
-	}
-
-	return s.value
+	n, _ := stream.First[LocalNode](context.Background(), s)
```

Comment:
Yep, `First` gets the first element of the stream or fails if context is cancelled or stream completes. 
`stream.Multicast` with `EmitLatest` will keep the last emitted element to give to any new subscriber, so calling `First` is effectively just picking whatever the last emitted item is.

To me `Only` doesn't make any sense here. What are its semantics if the stream has multiple elements? Would it assert that the stream must complete after the initial element? If so then we can't use it here as the stream never completes. We really just want the first element of the stream (e.g. the current latest).

Agreed regarding errors. I'll add a comment about the assumptions we rely on here.

Reply:
> stream.Multicast with EmitLatest will keep the last emitted element to give to any new subscriber, so calling First is effectively just picking whatever the last emitted item is.

I presume that `Last` would give the same result? By your explanation above I do find it a bit confusing why we'd use `First` if the intent is to get the `Last` element.

> To me Only doesn't make any sense here. What are its semantics if the stream has multiple elements? Would it assert that the stream must complete after the initial element? If so then we can't use it here as the stream never completes. We really just want the first element of the stream (e.g. the current latest).

The idea I was thinking about is that in cases like this, if there are somehow multiple elements in the stream, that is violating the expectations of code like this in the local node store. In such a scenario, there's likely some other bug that follows because local node store only ever picks one element, and the other one is ignored. Semantically at emit() or if necessary any time someone calls `Only()`, if there are multiple elements, it would return some error that could be raised up to the user (ideally actually to the developer) to announce clearly that somehow the expectation of "There's only one element" has been violated. That said, maybe this sort of suggestion is a little too paranoid. It sounds like the `EmitLatest` flag will guarantee that there's only ever one element, in which case unit tests should ensure that it's impossible to violate the expectation.

Reply:
> I presume that Last would give the same result? By your explanation above I do find it a bit confusing why we'd >
> use First if the intent is to get the Last element.

No, `Last` would block forever as the stream never completes. `Last` in the context of reactive streams means walk through the stream and give the last seen element before completion.

See: https://rxmarbles.com/#last

> The idea I was thinking about is that in cases like this, if there are somehow multiple elements in the stream, 
> that is violating the expectations of code like this in the local node store. In such a scenario, there's likely some > 
> other bug that follows because local node store only ever picks one element, and the other one is ignored. > 
> Semantically at emit() or if necessary any time someone calls Only(), if there are multiple elements, it would return
> some error that could be raised up to the user (ideally actually to the developer) to announce clearly that 
> somehow the expectation of "There's only one element" has been violated. That said, maybe this sort of
> suggestion is a little too paranoid. It sounds like the EmitLatest flag will guarantee that there's only ever one 
> element, in which case unit tests should ensure that it's impossible to violate the expectation.

`EmitLatest` means that when a new observer subscribes give it the last seen element, e.g. latest element.
I'm not sure where your impression that "there's only one element" comes from? `LocalNodeStore` emits essentially an infinite stream of new states and when subscribing we immediately give whatever was last emitted.

Reply:
> No, Last would block forever as the stream never completes. `Last` in the context of reactive streams means walk through the stream and give the last seen element before completion.

Shall we document this in the comment above `Last()`? The completion aspect of the semantics were not obvious to me from reading the code or code comments. For that matter, if the completion is a key aspect of the semantics, would it be more appropriate to name this `Final()` or `FinalItem()`? That said, I recognize that `Last` reflects the semantics described on rxmarbles. Personally I find "first" and "last" ambiguous when discussing streams given that it depends on the implementation details of the stream and exactly when the "first"/"last" call is made, but we can decide whether to clarify this through naming or code comments or developer education.

For that matter, the `return nil` semantics should probably also be documented. I guess that depends on whether the stream is Multicast or not, but if not then presumably only one caller can successfully call `Last()` and the subsequent ones will immediately return nil? Then if it's multicast, any number of callers can call `Last()` and each one will receive the most recent update prior to completion?

> `EmitLatest` means that when a new observer subscribes give it the last seen element. 

By itself, this makes sense to me and the code comments align well, I think this one was just on me to learn more about how it works. I don't have any better suggestions.

> I'm not sure where your impression that "there's only one element" comes from? LocalNodeStore emits essentially an infinite stream of new states and when subscribing we immediately give whatever was last emitted.

For the Local node, there is exactly one unique underlying object, the current state of the node. As that object is modified, it emits a stream of updates. As a programmer reasoning about the code, if I pick an arbitrary point in time and assume that there have been four updates, then it is tempting to visualize this as a list `[a, b, c, d]`. When I call `First()` on such a list, logically it would return the element at the start of the list, `a`. Theoretically this may represent stale data.  I suspect that the issue is that actually we don't track all updates `[a, b, c, d]` but rather we elide the updates. When the caller calls `First()`, `First` is not first as in "the beginning of all [previous and future] updates", but rather "the beginning of all future updates"? Or perhaps "The current state of the underlying object"? Maybe `MostRecent` or `MostRecentItem` would be more obvious?

Related, looking at rxmarbles for [first](https://rxmarbles.com/#first), I see:

![image](https://github.com/cilium/cilium/assets/1243336/04efa0d8-5619-4027-96e3-e0e3da5faf9a)

If I interpret time going from left to right, there are three objects {yellow, orange, green} and there are four updates, {1 (yellow), 2 (orange), 3 (green), 4 (yellow)}, then according to this diagram, `First` gives update 1. So that's not the most recent update for the yellow object. :thinking: 

Maybe these semantic changes to the meaning of `First()` are introduced specifically by the `EmitLatest()` flag(?) -- If so, that's surprising and should be called out in the code comment for `First()`.

Separately, contrasting to the local node case... if I consider updates for k8s Node or CiliumNode resources.. there are N unique underlying objects, each of which corresponds to some data stored in k8s, ultimately backed by another node in the cluster. Each of those objects may be updated independently. Given one stream for these resource updates, presumably if I call `First()`, it will provide an update for an arbitrary node. There's no guarantee which one it will be for. That implies to me that `First()` would be difficult to use for any code like for Nodes / CiliumNodes because inherently the stream is backed by multiple unique objects. It's not particularly meaningful to get the "first" or even "most recent" element for such a stream, and any such code probably *shouldn't* be trying to use `First()`, since it's missing updates for all the other nodes. By extension, if we were to try to make this code more programmer-safe, I could argue "It's a bug if someone tried to call `First()` for a stream where the updates correspond to multiple different underlying objects". A step even further would be "we should actively prevent the programmer from writing such bugs". I don't have a strong opinion where on this spectrum we are - prevent these bugs through developer education, through code comments/package API semantics, through review, through active code mechanisms.

Reply:
Related to my suggestions around `MostRecent()`, I think the key thing that I was missing is that `First()` immediately creates a brand new subscription in the multicast stream, _starting from the latest event at that point in time_, and therefore even if it's calling `First()` at that time, it's immediately already fast-forwarded to the current state when the call is made. `Latest` seems perhaps slightly more well-aligned for such semantics, but `MostRecent` can work or even `First()` if the code comments clearly describe that it's the First update that happens from this point forward, and ideally mention that when combined with `EmitLatest`, it is the most recent update that occurred in the past.

Reply:
The more I think about it, the more I'm inclined to agree that `First()` is the right naming for it, it would just help to have clearer comments around the semantics with respect to time, with respect to the way the function changes behaviour with different flags, and with respect to blocking behaviour (per the [other thread](https://github.com/cilium/cilium/pull/24866#discussion_r1218491983)).

Reply:
Thanks for digging through this! My approach with `pkg/stream` has been to just stick with the "reactive extensions" API as close as I can to "reuse" the established vocabulary (https://reactivex.io/documentation/observable.html). It gets confusing when the semantics of the stream are not that of a e.g. a simple container, but are somehow dependant on subscription time or consume the underlying resource.